commerce_coordinator/apps/commercetools/clients.py

```diff
@@ -1775,3 +1814,106 @@ def get_credit_variant_by_course_run(
                 f"Unable to find credit variant with course run key {course_run_key}",
             )
             raise err
+
+    @conditional_retry
+    def get_product_and_variant_by_course_run_key(
+        self, course_run_key: str
+    ) -> Tuple[ProductProjection | None, ProductVariant | None]:
+        """
+        Fetches a product and its variant by matching course run key with variant sku
+
+        Args:
+            course_run_key: course run key
+
+        Returns:
+            Tuple[ProductProjection, ProductVariant] if found, None otherwise.
+        """
+        try:
+            results = self.base_client.product_projections.search(
+                filter=f'variants.sku:"{course_run_key}"',
+                mark_matching_variants=True,
+                with_total=False,
+            ).results
+
+            for product in results:
+                all_variants = [product.master_variant, *product.variants]
+                for variant in all_variants:
+                    if variant.is_matching_variant and variant.sku == course_run_key:
+                        return product, variant
```

Comment:
can we loop on `*product.variants` directly since master_variant isn't needed to be looked for matching variant?